migrate to github actions

oliver-sanders opened the pull request:
Highlights:

* Swap TravisCI for GH Actions.
* Migrate flakytests back into the main test battery.
* Run unit tests X5 parallel *(number out of thin air)*.
* Run tests X8 parallel *(number out of thin air)*.

Impact:

* Start to finish for testing is now ~20mins.
* Tests may be "less" flaky, but all tests are slightly flaky.
* The test which seems to be failing the most often is now a unit test #3547
* Functional tests now seem to be passing fairly reliably.

Too much conversation a little more [gh] action:

* The action won't run on this PR because of GitHub's security model (actions don't get run until merged).
* So to see it in action raise a pull request against your own fork (no configuration required).
* Or alternatively head over to mine https://github.com/oliver-sanders/cylc/pull/17

Caveats:

* GH actions doesn't seem to allow you to selectively re-run individual jobs at the moment.

Todo (future work):

* [ ] Reinstate code coverage (after #3488).
* [ ] Cache dependencies where possible (see if we can get away from needing apt).
* [ ] Try to save the prove data files for flaky test logging (perhaps combine & report with a chron action).

<!-- The following requirements must be satisfied (with "[x]"). -->
<!-- Mark the PR as a Draft if all requirements are not yet satisfied. -->

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [ ] Does not contain off-topic changes (use other PRs for other changes).
- [x] Does not need tests (it is a test!).
- [x] No change log entry required (invisible to users).
- [x] No documentation update required.

oliver-sanders commented:
Note, if we do see occasional failures in the functional tests, which we should probably expect to, we can move back to the old system of automatically rerunning-failures in the action. This might not be ideal but it's effectively what we do anyway.

oliver-sanders commented:
> Wouldn't it be better to leave the flaky tests in-place?

I've merged them in one commit so we can revert this change at any time.

The reason I put them back is that we were still seeing regular failures in the non-flaky tests. I think we may have more flaky tests than regular tests by that measure. I also wasn't convinced by this approach in the first place. If a test only passes in a particular environment it's not really a valid test. Also in my early testing I ran the entire test battery (tests+flakytests) x8 parallel and it passed first time!

How do we handle flakytests in the future?

I've set up Travis to automatically re-run any failed tests once. This is a more efficient approach to what we've been doing manually anyway. With this the tests should pass first time. I think we *might* be able to cache the datafile prove generates using GH actions. That would give us logging on a test-by-test basis enabling us to track flaky tests.

oliver-sanders commented:
Can't seem to get mail working, I've tried `bsd-mail`. Has anyone configured mail on ubuntu?

(The tests check for `mail` by running `command -v mail`).

https://github.com/oliver-sanders/cylc/pull/17/checks?check_run_id=555937169#step:6:169

kinow commented:
Set up a `sendmail` on Ubuntu long ago, then started using `ssmtp` because it was easier for testing. And for mocking with Java used [FakeSMTP](http://nilhcem.com/FakeSMTP/). But honestly don't know much about how that's used by Cylc for testing.

wxtim commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,71 @@
+        run: |
+          pip install ."[all]"
+          pip install pytest-xdist
+      - name: Style
```
Perhaps I'm being fussy, but is there any way to refactor all the set-up above this point since it gets re-used every time?

oliver-sanders replied:
I did try to avoid this but:

* The setup for the two jobs is not actually the same.
* GitHub actions does not support yaml anchors.

oliver-sanders replied:
> is there a reason why unit-test is a hyphen and functional_test and underscore?

Because I fluffed up a rebase.

oliver-sanders replied:
> I'm imagining `unit-test: needs: common-setup`

The [GH Actions `needs`](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds) keyword is like a dependency in Cylc, it says a must run before b, but these are different jobs running in different environments, they don't share a container.

What we can do later (see todos) is to do the setup, cache dependencies then load that setup in future jobs. BUT this is easier said than done due to our need to install stuff via `apt-get`. Caching Python modules is easy enough, caching system packages is harder. I did some digging a while back and found a stackoverflow post where someone showed how to cache a particular dependency installed via apt-get, but there isn't a generic solution. Note GitHub has not provided any actions for working with apt-get.

wxtim replied:
None of this is really blocking - just me trying to gain a better understanding.

oliver-sanders replied:
No, it's a good point and one which irritates me!

oliver-sanders replied:
I've managed to consolidate things down to something pretty small now anyhoo.

oliver-sanders commented:
Good news, I got bsd-mail installed, bad news, the email tests are now failing, watch this space...

oliver-sanders commented:
Sticking the WIP label up whilst I try to get the mail tests working.

hjoliver commented:
Waiting on removal of WIP label (and email tests). 


My inclination was to agree with @kinow  on keeping flaky tests separate. @oliver-sanders reasoning doesn't sound entirely unreasonable, but still - is there any real downside to keeping them separate until they really prove themselves to be non-flaky?

> If a test only passes in a particular environment it's not really a valid test

I think its more about timing of concurrent(ish) events under load, which is just hard when functional-testing a workflow engine.  Presumably GH is handling the load better, but not sure if things will stay that way forever...

oliver-sanders commented:
Still can't get the mail tests to pass:

```
mail: Cannot open mailbox /var/mail/runner: Permission denied
```

That's in a ubuntu env with:

```
sudo apt-get install -y sqlite3 mailutils postfix
```

oliver-sanders commented:
I've failed at fixing the mail setup so rather than drag this out forever I've reverted the mail installation (but left the commit in for future reference) we can attempt to get this working in a future PR.

I've also reverted the merging of flakytests back with tests and given them their own job to run in.

Rebased and deconflicted.

oliver-sanders commented:
Blocked by #3588

hjoliver commented:
Unblocked.

oliver-sanders commented:
305cec2 should fix the last two tests (another DNS issue - see #3595)

hjoliver commented:
Seemingly weird things in the raw GH Actions logs for the functional tests on your fork:

- all(?) tests seem to be executed in all of batches 1-4 (but not 5)
- each test seems to be executed twice in one batch (which one varies), once as `tests/../blah.t` and again as `'./tests/../blah.t`

A randomly chosen example:
```bash
$ TEST=suite-state/06; for i in $(seq 1 5); do echo "# raw-log$i:"; grep $TEST raw-log$i | sed -e 's/^.*\] //'; done

# raw-log1:
tests/suite-state/06-format.t .................................................... ok    31140 ms ( 0.00 usr  0.02 sys + 21.84 cusr  2.27 csys = 24.13 CPU)
# raw-log2:
tests/suite-state/06-format.t .................................................. ok    57039 ms ( 0.00 usr  0.02 sys + 18.74 cusr  1.76 csys = 20.52 CPU)
# raw-log3:
tests/suite-state/06-format.t .................................................. ok    29995 ms ( 0.00 usr  0.01 sys + 20.81 cusr  2.19 csys = 23.01 CPU)
./tests/suite-state/06-format.t ................................................ ok    29943 ms ( 0.00 usr  0.01 sys + 18.94 cusr  1.85 csys = 20.80 CPU)
# raw-log4:
tests/suite-state/06-format.t .................................................. ok    27463 ms ( 0.01 usr  0.02 sys + 42.11 cusr  4.16 csys = 46.30 CPU)
# raw-log5:
```
Also, the time seems a lot longer than the advertised 20 min:
![shot](https://user-images.githubusercontent.com/2362137/81125294-d7aeba80-8f8b-11ea-9c54-387f199f9324.png)

hjoliver commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,83 @@
+          fi
+          etc/bin/run-functional-tests -j "${NPROC}" --state=save "${BASE}" || (
+            echo '=== Re-running Failed Tests'
+            CHUNK= etc/bin/run-functional-tests -j "${NPROC}" --state=save,failed
```
Ah, here's the error?
```suggestion
            CHUNK="${CHUNK}" etc/bin/run-functional-tests -j "${NPROC}" --state=save,failed
```

oliver-sanders replied:
Nah, but close. The prove wrapper was not intended to have to handle multiple test bases, will fiddle it a bit.

hjoliver commented:
The main test batches look better now, but the flaky test batch is running everything in `tests` as well, in the "re-running failed tests" bit.

oliver-sanders commented:
Dammit, fix things for tests/ break them for flakytests/.

I've now put in an exception to strip the `TESTS` if `--state=(!save)` is provided, that should learn it.

oliver-sanders commented:
Start to end 20 mins:

<img width="534" alt="Screenshot 2020-05-07 at 11 09 48" src="https://user-images.githubusercontent.com/16705946/81282658-620e2180-9053-11ea-94fc-b30f2d3a4d39.png">

https://www.youtube.com/watch?v=EKlFbMGTJVY
